### PR TITLE
enable deployment of whole service INT-style in dev mode

### PR DESCRIPTION
**Pull request by jim-minter**
Merge https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/RP-Config/pullrequest/4405588 at the same time as this PR.

Prerequisite of #1368.  This PR enables developers to deploy a representative RP environment in dev based on an existing shared development environment.  Current rough steps:

* `. ./env` and use hack/genconfig to generate a service configuration file for dev
* `make deploy` to deploy the RP (it will fail the first time)
* `make publish-image-aro` to push the aro image to the newly created ACR
* `go run ./cmd/aro mirror latest` to sync the relevant OpenShift release image to the newly created ACR
* manually fix up the certificates in the deployed key vaults as usual
* `make deploy` again and cross fingers

Once deploy has succeeded:
* `go run ./hack/tunnel <RP LB IP>` to create a tunnel from localhost:8443 to the RP.  The tunnel adds the client certs needed for `az aro` to work locally.
* `CLUSTER=xyz go run ./hack/create cluster`

Main changes:
* adds ACR and parent DNS zones to global deployment
* removes int and production RP_MODEs in favour of feature flags.  Idea is that in PROD, no feature flags are set (at least to begin with).  In INT, we'll use DisableDenyAssignments; in dev we'll use DisableDenyAssignments,DisableSignedCertificates,EnableDevelopmentAuthorizer,EnableARMHelper.
* EnableDevelopmentAuthorizer is a CA + subject name authenticator for use in this architecture.
* Add tool to auto-generate basic service config for developers.
* lots of clean-up of hard-coded values

### CODE REVIEW

**Comment by jim-minter** [comment]
@mikeandescavage fyi - contains a fair amount of deployment code implementation; this work will facilitate getting the new MAG environments built out with much less manual effort because it proves that we can build an INT environment from scratch.

File: pkg/env/armhelper.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+type armHelper struct {
+	log *logrus.Entry
+	env Interface
+
+	roleassignments authorization.RoleAssignmentsClient
+	applications    graphrbac.ApplicationsClient
+}
+
+func newARMHelper(env Interface, log *logrus.Entry) (ARMHelper, error) {
```
Comment (mjudeikis) [inline pkg/env/armhelper.go]: I would prefer to see order like `ctx, logger, evn, all other` but ou well :)

File: pkg/deploy/devconfig.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	clientCert, err := x509.ParseCertificate(client)
+	if err != nil {
+		return nil, err
+	}
+
+	sshPublicKey, err := ioutil.ReadFile(os.Getenv("HOME") + "/.ssh/id_rsa.pub")
```
Comment (mjudeikis) [inline pkg/deploy/devconfig.go:73]: Would be better of making it env? I personally dont use "this" key for work related stuff, and have separate one. But happy to compromise.

File: pkg/deploy/devconfig.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+				Configuration: &Configuration{
+					DatabaseAccountName:  to.StringPtr(os.Getenv("USER") + "-aro-" + _env.Location()),
+					KeyvaultPrefix:       to.StringPtr(os.Getenv("USER") + "-aro-" + _env.Location()),
+					StorageAccountDomain: to.StringPtr(os.Getenv("USER") + "aro" + _env.Location() + ".blob." + _env.Environment().StorageEndpointSuffix),
+				},
+			},
+		},
+		Configuration: &Configuration{
+			ACRResourceID:                to.StringPtr("/subscriptions/" + _env.SubscriptionID() + "/resourceGroups/" + os.Getenv("USER") + "-global/providers/Microsoft.ContainerRegistry/registries/" + os.Getenv("USER") + "aro"),
+			AdminAPICABundle:             to.StringPtr(string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: ca}))),
```
Comment (mjudeikis) [inline pkg/deploy/devconfig.go:93]: `pem.EncodeToMemory` nice :) didnt knew this existed :D

File: pkg/deploy/devconfig.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+			AdminAPIClientCertCommonName: to.StringPtr(clientCert.Subject.CommonName),
+			ARMAPICABundle:               to.StringPtr(string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: ca}))),
+			ARMAPIClientCertCommonName:   to.StringPtr(clientCert.Subject.CommonName),
+			ClusterMDSDConfigVersion:     to.StringPtr("2.3"),
```
Comment (mjudeikis) [inline pkg/deploy/devconfig.go:98]: ? hardconded, will this not change with audit logs, etc?

File: pkg/deploy/generator/resources_rp.go [hunk 5]
```diff
@@ -916,16 +915,16 @@ done
...
 }
 
+func (g *generator) rpParentDNSZone() *arm.Resource {
+	return g.dnsZone("[parameters('rpParentDomainName')]")
+}
+
+func (g *generator) rpClusterParentDNSZone() *arm.Resource {
+	return g.dnsZone("[parameters('clusterParentDomainName')]")
+}
+
```
Comment (mjudeikis) [inline pkg/deploy/generator/resources_rp.go:928]: :exploding_head: abstractions x 3 times :D

File: hack/tunnel/tunnel.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	log := utillog.GetLogger()
+
+	log.Printf("starting, git commit %s", version.GitCommit)
+
+	flag.Parse()
+
+	if err := run(context.Background(), log); err != nil {
+		log.Fatal(err)
+	}
+}
```
Comment (mjudeikis) [inline hack/tunnel/tunnel.go:148]: I would like to understand better why we need this tunneling thingy in the first place. Looks so simple that it raises my questions why it exits in the firs place.

File: pkg/frontend/middleware/headers.go [hunk 7]
```diff
@@ -8,20 +8,18 @@ import (
...
+		if strings.EqualFold(r.Header.Get("X-Ms-Return-Client-Request-Id"), "true") {
+			w.Header().Set("X-Ms-Client-Request-Id", r.Header.Get("X-Ms-Client-Request-Id"))
+		}
 
-			if isDevelopmentMode {
-				r.Header.Set("Referer", "https://localhost:8443"+r.URL.String())
-			}
+		if r.Header.Get("Referer") == "" { // used in development mode
+			r.Header.Set("Referer", "https://localhost:8443"+r.URL.String())
+		}
```
Comment (mjudeikis) [inline pkg/frontend/middleware/headers.go:21]: Remind me, why we have this? Something CLI related but dont remember why

File: pkg/util/cluster/cluster.go [hunk 8]
```diff
@@ -417,6 +419,11 @@ func (c *Cluster) createCluster(ctx context.Context, vnetResourceGroup, clusterN
 	}
 
 	if c.env.IsDevelopmentMode() {
+		err := c.registerSubscription(ctx)
```
Comment (mjudeikis) [inline pkg/util/cluster/cluster.go:422]: +1

**Review by mjudeikis — COMMENTED** [review]
I think this is ok, just want to test it locally and I think we can merge

**Comment by mjudeikis** [comment]
Dev mode needs `acrReplicaDisabled: true` due to
```
FATA[2021-04-14T16:25:18+01:00]cmd/aro/main.go:75 main.main() Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details." Details=[{"code":"BadRequest","message":"{\r\n  \"error\": {\r\n    \"code\": \"HomeReplicationCannotBeOperated\",\r\n    \"message\": \"Registry replication must specify a region different from the home registry in westeurope. For more information, please visit https://aka.ms/acr/geo-replication\"\r\n  },\r\n  \"status\": \"Failed\"\r\n}"}]```

**Comment by mjudeikis** [comment]
```       "keyvaultPrefix": {
            "type": "string",
            "maxLength": 20
        },```
but we generate it long :) `keyvaultPrefix: mjudeiki-aro-westeurope`

File: pkg/deploy/devconfig.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+	return &Config{
+		RPs: []RPConfig{
+			{
+				Location:            _env.Location(),
+				SubscriptionID:      _env.SubscriptionID(),
+				RPResourceGroupName: os.Getenv("USER") + "-aro-" + _env.Location(),
+				Configuration: &Configuration{
+					DatabaseAccountName:  to.StringPtr(os.Getenv("USER") + "-aro-" + _env.Location()),
+					KeyvaultPrefix:       to.StringPtr(os.Getenv("USER") + "-aro-" + _env.Location()),
```
Comment (mjudeikis) [inline pkg/deploy/devconfig.go:86]: This has limit of 20 in the arm template. And any if you are not `Jim` you most probably will go over this limit :) I would change it to env variable fully and let people to control it.

**Comment by mjudeikis** [comment]
Would be really good to have clean-up helper for this too... As I can see how this GC can become a problem.

File: pkg/util/clientauthorizer/admin.go [hunk 10]
```diff
@@ -17,10 +17,6 @@ import (
 // contain a valid client certificate signed by `caBundlePath` and
 // the client certificate's CommonName equals `clientCertCommonName`.
 func NewAdmin(log *logrus.Entry, caBundlePath, clientCertCommonName string) (ClientAuthorizer, error) {
-	if clientCertCommonName == "" {
```
Comment (jim-minter) [inline pkg/util/clientauthorizer/admin.go:20]: make clear what the code path is that causes this to "fail" closed when no configuration is provided

File: Makefile [hunk 11]
```diff
@@ -24,6 +24,9 @@ clean:
 client: generate
 	hack/build-client.sh 2020-04-30 2021-01-31-preview
 
+deploy:
```
Comment (jim-minter) [inline Makefile:27]: add a note to say that this is hacky and hopefully temporary and open for improvement

Comment (jim-minter) [inline Makefile:27]: add makefile target for `go run ./hack/genconfig`

File: deploy/rp-production-parameters.json [hunk 12]
```diff
@@ -44,9 +44,6 @@
         "mdmFrontendUrl": {
             "value": ""
         },
-        "mdsdConfigVersion": {
```
Comment (jim-minter) [inline deploy/rp-production-parameters.json:47]: this change requires the RP-Config to be changed at the same time - prepare a PR for this.

File: pkg/deploy/config.go [hunk 13]
```diff
@@ -22,7 +22,7 @@ type Config struct {
 type RPConfig struct {
 	Location            string         `json:"location,omitempty"`
 	SubscriptionID      string         `json:"subscriptionId,omitempty"`
-	RPResourceGroupName string         `json:"resourceGroupName,omitempty"`
+	RPResourceGroupName string         `json:"rpResourceGroupName,omitempty"`
```
Comment (jim-minter) [inline pkg/deploy/config.go:28]: another RP-Config change needed

File: pkg/cluster/delete.go [hunk 14]
```diff
@@ -303,7 +303,7 @@ func (m *manager) Delete(ctx context.Context) error {
 		return err
 	}
 
-	if !m.env.IsDevelopmentMode() {
```
Comment (jim-minter) [inline pkg/cluster/delete.go:306]: maybe worth renaming IsDevelopmentMode to something like "is on your laptop or in CI mode"?

File: pkg/env/prod.go [hunk 15]
```diff
@@ -167,7 +167,21 @@ func newProd(ctx context.Context, log *logrus.Entry) (*prod, error) {
 }
 
 func (p *prod) InitializeAuthorizers() error {
-	p.armClientAuthorizer = clientauthorizer.NewARM(p.log, p.Core)
+	if !p.FeatureIsSet(FeatureEnableDevelopmentAuthorizer) {
+		p.armClientAuthorizer = clientauthorizer.NewARM(p.log, p.Core)
+
+	} else {
+		armClientAuthorizer, err := clientauthorizer.NewAdmin(
```
Comment (jim-minter) [inline pkg/env/prod.go:180]: maybe renaming NewAdmin to "CA + SAN" or something might be clearer

File: cmd/aro/rp.go [hunk 16]
```diff
@@ -45,7 +45,6 @@ func rp(ctx context.Context, log, audit *logrus.Entry) error {
 	} else {
 		keys = []string{
 			"ACR_RESOURCE_ID",
-			"ADMIN_API_CLIENT_CERT_COMMON_NAME",
```
Comment (jim-minter) [inline cmd/aro/rp.go:48]: could consider entirely dropping this commit given that we are setting up the ARM and admin authorizers against a developer CA

File: pkg/env/armhelper.go [hunk 17]
```diff
@@ -0,0 +1,105 @@
+package env
```
Comment (jim-minter) [inline pkg/env/armhelper.go:1]: this commit got messed up and the feature flag is lost and the FPClientID changes seem to have got inserted

Comment (jim-minter) [reply pkg/env/armhelper.go:1]: documentation needed for armhelper.

File: pkg/env/env.go [hunk 18]
```diff
@@ -27,6 +27,7 @@ const (
 	FeatureDisableDenyAssignments Feature = iota
 	FeatureDisableSignedCertificates
 	FeatureEnableDevelopmentAuthorizer
+	FeatureRequireD2sV3Workers
```
Comment (jim-minter) [inline pkg/env/env.go:30]: check: the change to dev config generation which adds FeatureRequireD2sV3Workers should also be in this commit

File: docs/prepare-a-shared-rp-development-environment.md [hunk 19]
```diff
@@ -78,16 +78,24 @@ locations.
 
 1. Create an AAD application which will fake up the ARM layer:
```
Comment (jim-minter) [inline docs/prepare-a-shared-rp-development-environment.md:79]: add documentation about feature flags - 
subscription feature flags
RP feature flag environment variable
cluster document feature flags

File: pkg/deploy/generator/resources_rp.go (hunk 5 above)
Comment (jim-minter) [reply pkg/deploy/generator/resources_rp.go:928]: ech, it's not perfect but I think it's good enough.  No-one will feel bad about changing it.

**Comment by jim-minter** [comment]
> Dev mode needs acrReplicaDisabled: true due to

@mjudeikis fixed in f3df81328f50d4ad2de6d744c7ab57056025d196

**Comment by jim-minter** [comment]
> Would be really good to have clean-up helper for this too... As I can see how this GC can become a problem.
If not this PR - Issue/story

@mjudeikis leaving this for someone else, feel free to open a story.

File: deploy/rp-production-parameters.json (hunk 12 above)
Comment (jim-minter) [reply deploy/rp-production-parameters.json:47]: https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/RP-Config/pullrequest/4405588

File: hack/tunnel/tunnel.go [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+				defer func() {
+					_ = c2.CloseWrite()
+				}()
+				_, _ = io.Copy(c2, c)
+			}()
+
+			defer func() {
+				_ = c.(*tls.Conn).CloseWrite()
+			}()
+			_, _ = io.Copy(c, c2)
```
Comment (jim-minter) [inline hack/tunnel/tunnel.go:146]:
```suggestion
			func() {
				defer func() {
					_ = c.(*tls.Conn).CloseWrite()
				}()
				_, _ = io.Copy(c, c2)
			}()
```

**Review by mjudeikis — APPROVED** [review]

**Pull request merged**